File: 프로그래머스/unrated/178871. 달리기 경주/달리기 경주.java
import java.util.*;
import java.util.stream.*;

class Solution {
    public String[] solution(String[] players, String[] callings) {
        List<String> playerList = new ArrayList<>(Arrays.asList(players));
        Map<String, Integer> playerIndexMap = IntStream.range(0, players.length)
                .boxed()
                .collect(Collectors.toMap(i -> players[i], i -> i));

        Arrays.stream(callings)
              .forEach(calling -> {
                  int index = playerIndexMap.get(calling);
                  if (index > 0) {
                      Collections.swap(playerList, index, index - 1);
                      playerIndexMap.put(calling, index - 1);
                      playerIndexMap.put(playerList.get(index), index);
                  }
              });

        return playerList.toArray(new String[0]);
    }
}

File: 프로그래머스/unrated/178871. 달리기 경주/달리기 경주.kt
import java.util.*;

class Solution {
    fun solution(players: Array<String>, callings: Array<String>): Array<String> {
        val playerList = players.toMutableList()
        val playerIndexMap: MutableMap<String, Int> = players.withIndex().associate { it.value to it.index }.toMutableMap()

        callings.forEach { calling ->
            val index = playerIndexMap[calling] ?: return@forEach
            if (index > 0) {
                Collections.swap(playerList, index, index - 1)
                playerIndexMap[playerList[index - 1]] = index - 1
                playerIndexMap[playerList[index]] = index
            }
        }
        return playerList.toTypedArray()
    }
}
